### DISCUSSION TITLE
CSLA with Serverless / AWS Lambda

### DISCUSSION THREAD

**Discussion by jonnysmith1981** [MarimerLLC/csla#3723] [Questions] [answered]
Does anyone have any examples or real world experience of using CSLA with serverless technologies, most notably AWS Lambda?

**Comment by rockfordlhotka** [comment]
I don't know about Lambda, but I have had a couple people talk about using CSLA within Azure Functions over the years. It used to be a bit tricky with the very first implementation of Functions due to the way Microsoft hosted the code, but Functions work differently now and afaik there's no problem.

**Comment by swegele** [comment, marked as answer]
I've been messing with Azure Functions for a few weeks.  It builds and compiles like any other c# project.  This technology feels like a stripped down one-way console app (meaning no ui input).  I am able to configure and initialize DI services in Program.cs and to inject those services into my Function class just like any other c# class. 

So there is nothing about serverless that would cause trouble with CSLA.  But obviously you'd be using a pretty small subset of what CSLA can offer.  Throw out the DataPortal.  Throw out UI databinding.  Probably the only real feature left of value would be business rules?  I'm not sure of the value proposition there.

P.S.  The inputs/bindings are finicky for functions.  The documentation lagging and inconsistent on Azure Functions v2x and 4x (probably because things are changing so rapidly as they iterate).  The examples are often mixed between the older versions and the newer ones or incomplete not showing what is in the appsettings file.  So you are left to "plug and play-with-it" experience.  I'm sure it will catch up eventually.

**Reply by hurcane** [reply]
I think is a case of the classic consultant answer, "It depends." 

I think Business Rules is one of the strongest features of CSLA. If your Azure function needs to validate data and persist that data, then CSLA is something to consider. If the business classes can also be used for the same business use case, but via UI, then that is an even stronger case to use CSLA. I've found it indispensable for having common validation logic between human data entry and automated data entry (via an API).

If your function doesn't need to evaluate complex business requirements and update data, then CSLA might not be the right fit.
